File: library/src/main/java/pw/bits/weisper/library/WeiboModel.java
package pw.bits.weisper.library;

import java.util.List;

import pw.bits.weisper.library.bean.Status;
import pw.bits.weisper.library.bean.StatusResponse;
import pw.bits.weisper.library.bean.User;
import rx.Observable;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * Created by rzh on 16/3/14.
 */
public class WeiboModel {
    private static WeiboService service;

    public static void init(WeiboService service) {
        WeiboModel.service = service;
    }

    public static Observable<StatusResponse> statusesHomeTimeline(long since_id, long max_id) {
        return service.statusesHomeTimeline(20, since_id, max_id)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static Observable<StatusResponse> statusesUserTimeLine(String screen_name, long since_id, long max_id) {
        return service.statusesUserTimeLine(20, since_id, max_id, screen_name)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static Observable<StatusResponse> searchTopics(String topic) {
        return service.searchTopics(50, topic)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static Observable<User> usersShow(String screen_name) {
        return service.usersShow(screen_name, null)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static Observable<User> usersShow(Long uid) {
        return service.usersShow(null, uid)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static Observable<Status> statusesRepost(Long id, String status, Boolean is_comment) {
        return service.statusesRepost(id, status, is_comment ? 1 : 0)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    public static Observable<StatusResponse> statusesShowBatch(List<Long> ids) {
        String string = "";
        for (int i = 0; i < ids.size() - 1; i++) {
            string += ids.get(i) + ",";
        }
        if (ids.size() > 0) {
            string += ids.get(ids.size() - 1);
        }
        return service.statusesShowBatch(string)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }
}

File: library/src/main/kotlin/pw/bits/weisper/library/WeiboModel.kt
package pw.bits.weisper.library

import com.facebook.stetho.okhttp3.StethoInterceptor
import com.google.gson.GsonBuilder
import com.orhanobut.hawk.Hawk
import okhttp3.Interceptor

import okhttp3.OkHttpClient
import okhttp3.Response
import pw.bits.weisper.library.bean.Status
import pw.bits.weisper.library.bean.Statuses
import pw.bits.weisper.library.bean.User
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import rx.Observable
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * Created by rzh on 16/3/14.
 */
object WeiboModel {
    private val client = OkHttpClient.Builder().addInterceptor(interceptor).addNetworkInterceptor(StethoInterceptor()).build()

    private val converter = GsonConverterFactory.create(GsonBuilder().setDateFormat("EEE MMM dd HH:mm:ss z yyyy").create());

    private val adapter = RxJavaCallAdapterFactory.create();

    private val retrofit = Retrofit.Builder().addConverterFactory(converter).addCallAdapterFactory(adapter).client(client).baseUrl("https://api.weibo.com/2/").build()

    private val service = retrofit.create(WeiboService::class.java)

    object interceptor : Interceptor {
        override fun intercept(chain: Interceptor.Chain): Response {
            val original = chain.request()
            val url = original.url().newBuilder().addQueryParameter("access_token", Hawk.get("access-token", "")).build()
            val request = original.newBuilder().url(url).method(original.method(), original.body()).build()
            return chain.proceed(request)
        }
    }

    fun statusesHomeTimeline(since_id: Long, max_id: Long): Observable<Statuses> {
        return service.statusesHomeTimeline(since_id, max_id).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun statusesUserTimeLine(screen_name: String, since_id: Long, max_id: Long): Observable<Statuses> {
        return service.statusesUserTimeLine(since_id, max_id, screen_name).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun searchTopics(topic: String): Observable<Statuses> {
        return service.searchTopics(topic).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun usersShow(screen_name: String): Observable<User> {
        return service.usersShow(screen_name).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun usersShow(uid: Long): Observable<User> {
        return service.usersShow(uid).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun statusesRepost(id: Long, status: String, is_comment: Boolean): Observable<Status> {
        return service.statusesRepost(id, status, if (is_comment) 1 else 0).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }

    fun statusesShowBatch(ids: List<Long>): Observable<Statuses> {
        return service.statusesShowBatch(ids.joinToString(",")).subscribeOn(Schedulers.io()).observeOn(AndroidSchedulers.mainThread())
    }
}
